src/network/networkd-dhcp4.c

```diff
@@ -583,6 +586,48 @@ static int dhcp4_set_hostname(Link *link) {
         return sd_dhcp_client_set_hostname(link->dhcp_client, hn);
 }
 
+static bool check_promote_secondaries(const char *ifname) {
+        char *promote_secondaries_sysctl = NULL, *promote_secondaries_path = NULL;
+
+        promote_secondaries_path = strjoina("net/ipv4/conf/", ifname, "/promote_secondaries");
+        if (sysctl_read(promote_secondaries_path, &promote_secondaries_sysctl) >= 0) {
+                truncate_nl(promote_secondaries_sysctl);
+                return parse_boolean(promote_secondaries_sysctl);
+        }
+
+        return false;
+}
+
+// dhcp4_check_promote_secondaries will ensure this interface has
+// the "promote_secondaries" option in the kernel set. If this sysctl
+// is not set DHCP will work only as long as the IP address does not
+// changes between leases. The kernel will remove all secondary IP
+// addresses of an interface otherwise. The way systemd-network works
+// is that the new IP of a lease is added as a secondary IP and when
+// the primary one expires it relies on the kernel to promote the
+// secondary IP. See also https://github.com/systemd/systemd/issues/7163
```

Comment:
no C++ // comments, please. We use C comments, see CODING_STYLE.

This comment contains what really should be in the man page btw: the fact that secondary IP addresses are promoted to primary ones, when the primary ones expire